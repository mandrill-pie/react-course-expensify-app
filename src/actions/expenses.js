import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
	type: 'ADD_EXPENSE',
	expense
});

// Async action
export const startAddExpense = (expenseData = {}) => {
	return (dispatch) => {
		// Destructure received data (expenseData)
		const {
			description = '',
			note = '',
			amount = 0,
			createdAt = 0
		} = expenseData;

		// Save to db
		const expense = {
			description,
			note,
			amount,
			createdAt
		};
		return database.ref('expenses')
			.push(expense)
			.then((ref) => {
				// Update the Redux Store
				dispatch(addExpense({
					id: ref.key,
					...expense
				}));
			});
	};
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
});

// Async action
export const startRemoveExpense = ({ id }) => {
	return (dispatch) => {
		// remove expense from DB
		return database.ref(`expenses/${id}`).remove().then(() => {
			// remove expense from state
			dispatch(removeExpense({ id }));
		});
	};
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});

// Async action
export const startEditExpense = (id, updates) => {
	return (dispatch) => {
		// Update expense in DB
		return database.ref(`expenses/${id}`)
			.update(updates)
			.then(() => {
				// Update the Redux Store
				dispatch(editExpense(id, updates));	
			})
			.catch((error) => {
				console.log('Error: ', error);
			});
	};
};

// SET_EXPENSES (get expenses from DB)
export const setExpenses = (expenses) => ({
	type: 'SET_EXPENSES',
	expenses
});

// Async action
export const startSetExpenses = () => {
	return (dispatch) => {
		// Get expenses from DB
		return database.ref('expenses')
			.once('value')
			.then((snapshot) => {
				const expenses = [];
				snapshot.forEach((childSnapshot) => {
					expenses.push({
						...childSnapshot.val(),
						id: childSnapshot.key,
					});
				});

				// Update the Redux Store
				dispatch(setExpenses(expenses));	
			})
			.catch((error) => {
				console.log('Error: ', error);
			});
	};
};