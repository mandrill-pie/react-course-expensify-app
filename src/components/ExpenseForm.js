import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
	// need a constructor to get access to props
	constructor(props) {
		super(props);

		this.state = {
			description: props.expense ? props.expense.description : '',
			amount: props.expense ? props.expense.amount : '',
			createdAt: props.expense ? moment(props.expense.createdAt): moment(),
			note: props.expense ? props.expense.note : '',
			calendarFocused: false,
			error: false,
		};
	}

	onDescriptionChange = (e) => {
		const description = e.target.value;
		this.setState(() => ({ description }));
	}
	onNoteChange = (e) => {
		const note = e.target.value;
		this.setState(() => ({ note }));
	}
	onAmountChange = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,10}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	}
	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	}
	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }));	
	}
	onSubmit = (e) => {
		e.preventDefault();

		if (!this.state.description || !this.state.amount) {
			this.setState(() => ({ error: 'Please provide a description and amount!' }));
		} else {
			this.setState(() => ({ error: false }));
			// call props.onSubmit() that gets passed to this component
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount),
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note,
			});
		}

	}
	render() {
		return (
			<form className="form" onSubmit={this.onSubmit}>
				{
					this.state.error &&
					<p className="form__error">{this.state.error}</p>
				}
				<input
					className="input-text"
					type="text"
					placeholder="Description"
					autoFocus
					value={this.state.description}
					onChange={this.onDescriptionChange} />
				<input
					className="input-text"
					type="text" // number
					placeholder="Amount"
					value={this.state.amount}
					onChange={this.onAmountChange} />
				<SingleDatePicker
					date={this.state.createdAt}
					onDateChange={this.onDateChange}
					focused={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					numberOfMonths={1}
					isOutsideRange={() => false} />
				<textarea
					className="textarea"
					placeholder="Add a note"
					onChange={this.onNoteChange}
					value={this.state.note}>
				</textarea>

				<input className="btn" type="submit" value="Save expense" />
			</form>
		);
	}
}