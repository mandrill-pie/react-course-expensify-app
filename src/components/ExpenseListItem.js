import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ExpenseListItem = ({id, description, amount, createdAt}) => (
	<tr>
		<td>
			{id}
		</td>
		<td>
			{description}
		</td>
		<td>{amount}</td>
		<td>{moment(createdAt).format('YYYY-MM-DD')}</td>
		<td>
			<Link to={`/edit/${id}`}>Edit</Link>
		</td>
	</tr>
);

export default ExpenseListItem;