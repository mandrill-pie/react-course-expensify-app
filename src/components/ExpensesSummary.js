import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export class ExpensesSummary extends React.Component {

	render = () => (
		<div>
			Viewing	{this.props.expenseCount}	{this.props.expenseCount > 1 ? 'expenses' : 'expense'} totalling {this.props.expensesTotal}
		</div>
	)
}

const mapStateToProps = (state) => ({
	expenseCount: selectExpenses(state.expenses, state.filters).length,
	expensesTotal: selectExpensesTotal(state.expenses)
});

export default connect(mapStateToProps)(ExpensesSummary);