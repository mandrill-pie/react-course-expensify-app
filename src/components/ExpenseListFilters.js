import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setFilterText, sort, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			calendarFocused: null,
		};
	}

	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	}

	onFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }));
	}

	onTextChange = (e) => {
		this.props.setFilterText(e.target.value);
	}

	onSortChange = (e) => {
		this.props.sort(e.target.value);
	}

	render = () => (
		<div>
			<input
				type="text"
				value={this.props.filters.text}
				onChange={this.onTextChange} />
			<select 
				value={this.props.filters.sortBy}
				onChange={this.onSortChange}>
				<option value="date">Date</option>
				<option value="amount">Amount</option>
			</select>
			<DateRangePicker
				startDate={this.props.filters.startDate}
				endDate={this.props.filters.endDate}
				onDatesChange={this.onDatesChange}
				focusedInput={this.state.calendarFocused}
				onFocusChange={this.onFocusChange}
				showClearDates={true}
				numberOfMonths={1}
				isOutsideRange={() => false} />
		</div>
	);
}

const mapStateToProps = (state) => ({
	filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate)),
	setFilterText: (text) => dispatch(setFilterText(text)),
	sort: (value) => dispatch(sort(value)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);