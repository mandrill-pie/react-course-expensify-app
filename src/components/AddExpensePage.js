import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
	// no constructor needed to get access to props!

	onSubmit = (expense) => {
		// dispatch by using addExpense (defined in mapDispatchToProps)
		// workaround for testing, used instead of:
		// props.dispatch(addExpense(expense));
		this.props.addExpense(expense);
		// redirect
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<ExpenseForm
					onSubmit={this.onSubmit}
				/>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({
	// Give AddExpensePage component access to addExpense()
	addExpense: (expense) => dispatch(addExpense(expense))
});

// give AddExpensePage access to props
// so it can use props.dispatch()
// do not pass a mapStateToProps() function
// but do pass a mapDispatchToProps() (used in testing)
export default connect(false, mapDispatchToProps)(AddExpensePage);