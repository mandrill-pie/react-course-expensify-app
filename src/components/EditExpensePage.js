import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {

	onSubmit = (updates) => {
		// dispatch
		this.props.startEditExpense(this.props.expense.id, updates);
		// redirect
		this.props.history.push('/');
	}

	onRemove = () => {
		this.props.startRemoveExpense({ id: this.props.expense.id });
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<ExpenseForm
					expense={this.props.expense}
					onSubmit={this.onSubmit}
				/>
				<button onClick={this.onRemove}>Remove</button>
			</div>
		);
	}
}

const mapStateToProps = (state, props) => ({
	// populate props.expense with the required expense (based on id)
	expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
	// Give parent component (EditExpensePage) access to editExpense()
	// Needs access to props (second argument)! <- WHY?
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
});

// give EditExpensePage access to props
// so it can use props.dispatch()
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);