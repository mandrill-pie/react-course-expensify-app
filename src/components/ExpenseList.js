import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export class ExpenseList extends React.Component {
	constructor(props) {
		super(props);
	}

	render = () => {
		if (this.props.expenses.length === 0) {
			return <p>No expenses found</p>;
		}	else {
			return (
				<table>
					<tbody>
						{
							this.props.expenses.map((expense) => (
								<ExpenseListItem key={expense.id} {...expense} />
							))
						}
					</tbody>
				</table>
			);
		}
	}
}

const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpenseList);