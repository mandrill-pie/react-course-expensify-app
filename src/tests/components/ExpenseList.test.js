import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('ExpenseList - render list of expenses', () => {
	const wrapper = shallow(<ExpenseList expenses={expenses} />);
	expect(wrapper).toMatchSnapshot();
});

test('ExpenseList - render with no expenses', () => {
	const wrapper = shallow(<ExpenseList expenses={[]} />);
	expect(wrapper).toMatchSnapshot();
});