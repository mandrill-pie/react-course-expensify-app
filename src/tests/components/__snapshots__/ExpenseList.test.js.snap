// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList - render list of expenses 1`] = `
<table>
  <tbody>
    <ExpenseListItem
      amount={2.76}
      createdAt={1505145600000}
      description="Gum"
      id={1}
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={1.95}
      createdAt={1487779200000}
      description="Mints"
      id={2}
      key="2"
      note="Some green mints"
    />
    <ExpenseListItem
      amount={143.16}
      createdAt={1507132800000}
      description="Rented car"
      id={3}
      key="3"
      note="From my trip to New Zeeland"
    />
    <ExpenseListItem
      amount={14.95}
      createdAt={1509811200000}
      description="Gummy bears"
      id={4}
      key="4"
      note="A whole box"
    />
  </tbody>
</table>
`;

exports[`ExpenseList - render with no expenses 1`] = `
<p>
  No expenses found
</p>
`;
